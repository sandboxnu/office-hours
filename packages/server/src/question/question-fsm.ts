import {
  ClosedQuestionStatus,
  LimboQuestionStatus,
  OpenQuestionStatus,
  QuestionStatus,
  Role,
} from '@koh/common';

interface AllowableTransitions {
  student?: QuestionStatus[];
  ta?: QuestionStatus[];
}

const QUEUE_TRANSITIONS: AllowableTransitions = {
  ta: [OpenQuestionStatus.Helping, LimboQuestionStatus.TADeleted],
  student: [ClosedQuestionStatus.ConfirmedDeleted],
};

const QUESTION_STATES: Record<QuestionStatus, AllowableTransitions> = {
  [OpenQuestionStatus.Drafting]: {
    student: [OpenQuestionStatus.Queued, ClosedQuestionStatus.ConfirmedDeleted],
    ta: [OpenQuestionStatus.Helping, ClosedQuestionStatus.DeletedDraft],
  },
  [OpenQuestionStatus.Queued]: QUEUE_TRANSITIONS,
  [OpenQuestionStatus.PriorityQueued]: QUEUE_TRANSITIONS,
  [OpenQuestionStatus.Helping]: {
    ta: [
      LimboQuestionStatus.CantFind,
      LimboQuestionStatus.ReQueueing,
      ClosedQuestionStatus.Resolved,
      LimboQuestionStatus.TADeleted,
    ],
    student: [ClosedQuestionStatus.ConfirmedDeleted],
  },
  [LimboQuestionStatus.CantFind]: {
    student: [
      OpenQuestionStatus.PriorityQueued,
      ClosedQuestionStatus.ConfirmedDeleted,
    ],
  },
  [LimboQuestionStatus.ReQueueing]: {
    student: [
      OpenQuestionStatus.PriorityQueued,
      OpenQuestionStatus.Queued,
      ClosedQuestionStatus.ConfirmedDeleted,
    ],
  },
  [LimboQuestionStatus.TADeleted]: {
    student: [ClosedQuestionStatus.ConfirmedDeleted],
  },
  [ClosedQuestionStatus.Resolved]: {},
  [ClosedQuestionStatus.ConfirmedDeleted]: {},
  [ClosedQuestionStatus.Stale]: {},
  [ClosedQuestionStatus.DeletedDraft]: {},
};

export function canChangeQuestionStatus(
  oldStatus: QuestionStatus,
  goalStatus: QuestionStatus,
  role: Role,
): boolean {
  return (
    oldStatus === goalStatus ||
    QUESTION_STATES[oldStatus][role]?.includes(goalStatus)
  );
}
