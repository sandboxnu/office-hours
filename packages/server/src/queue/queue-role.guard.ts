import { Injectable, UnauthorizedException } from '@nestjs/common';
import { UserModel } from '../profile/user.entity';
import { RolesGuard } from '../guards/role.guard';
import { QueueModel } from './queue.entity';

@Injectable()
export class QueueRolesGuard extends RolesGuard {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  async setupData(
    request: any,
  ): Promise<{ courseId: number; user: UserModel }> {
    const queue = await QueueModel.findOne(request.params.queueId);
    const courseId = queue?.courseId;
    const user = await UserModel.findOne(request.user.userId, {
      relations: ['courses'],
    });

    return { courseId, user };
  }
}
