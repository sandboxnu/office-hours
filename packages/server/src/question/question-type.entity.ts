import {
  BaseEntity,
  Column,
  Entity,
  ManyToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { QuestionModel } from './question.entity';

@Entity('question_type_model')
export class QuestionTypeModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ nullable: true })
  cid: number;

  @Column({ type: 'text', nullable: true })
  name: string;

  @Column({ type: 'text', nullable: true })
  color: string;

  @ManyToMany(() => QuestionModel, (question) => question.questionTypes)
  questions: QuestionModel[];
}
