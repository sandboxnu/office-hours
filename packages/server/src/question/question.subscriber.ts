import {
  EventSubscriber,
  EntitySubscriberInterface,
  Connection,
  UpdateEvent,
  InsertEvent,
} from 'typeorm';
import {
  NotificationService,
  NotifMsgs,
} from '../notification/notification.service';
import { QuestionModel } from './question.entity';
import { ClosedQuestionStatus } from '@template/common';
import { QueueSSEService } from 'queue/queue-sse.service';

@EventSubscriber()
export class QuestionSubscriber
  implements EntitySubscriberInterface<QuestionModel> {
  private notifService: NotificationService;
  private queueSSEService: QueueSSEService;
  constructor(
    connection: Connection,
    notifService: NotificationService,
    queueSSEService: QueueSSEService,
  ) {
    this.notifService = notifService;
    this.queueSSEService = queueSSEService;
    connection.subscribers.push(this);
  }

  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  listenTo() {
    return QuestionModel;
  }

  async afterUpdate(event: UpdateEvent<QuestionModel>): Promise<void> {
    // Send all listening clients an update
    await this.queueSSEService.updateQuestions(event.entity.queueId);

    // Send push notification to students when they are hit 3rd in line
    // if status updated to closed
    if (
      event.updatedColumns.find((c) => c.propertyName === 'status') &&
      event.entity.status in ClosedQuestionStatus
    ) {
      // get 3rd in queue before and after this update
      const previousThird = await QuestionModel.openInQueue(
        event.entity.queueId,
      )
        .offset(2)
        .getOne();
      const third = await QuestionModel.openInQueue(event.entity.queueId)
        .setQueryRunner(event.queryRunner) // Run in same transaction as the update
        .offset(2)
        .getOne();
      if (previousThird?.id !== third?.id) {
        const { creatorId } = third;
        this.notifService.notifyUser(creatorId, NotifMsgs.queue.THIRD_PLACE);
      }
    }
  }

  async afterInsert(event: InsertEvent<QuestionModel>): Promise<void> {
    // Send all listening clients an update
    await this.queueSSEService.updateQuestions(event.entity.queueId);
  }
}
