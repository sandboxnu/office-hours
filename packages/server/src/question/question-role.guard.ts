import {
  Injectable,
  UnauthorizedException,
  NotFoundException,
} from '@nestjs/common';
import { UserModel } from '../profile/user.entity';
import { QuestionModel } from './question.entity';
import { QueueModel } from '../queue/queue.entity';
import { RolesGuard } from '../guards/role.guard';

@Injectable()
export class QuestionRolesGuard extends RolesGuard {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  async setupData(
    request: any,
  ): Promise<{ courseId: number; user: UserModel }> {
    let queueId = request.params.queueId;

    //specific case when we are posting a new question to the queue
    if (request.params.questionId && queueId) {
      const question = await QuestionModel.findOne(request.params.questionId);
      queueId = question.queueId;
    }

    const queue = await QueueModel.findOne(queueId);

    // You cannot interact with a question in a nonexistent queue
    if (!queue) {
      throw new NotFoundException('This queue does not exist!');
    }
    const courseId = queue.courseId;
    const user = await UserModel.findOne(request.user.userId, {
      relations: ['courses'],
    });

    return { courseId, user };
  }
}
