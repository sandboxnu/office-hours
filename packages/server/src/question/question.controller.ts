import {
  ClosedQuestionStatus,
  CreateQuestionParams,
  CreateQuestionResponse,
  ERROR_MESSAGES,
  GetQuestionResponse,
  GroupQuestionsParams,
  LimboQuestionStatus,
  OpenQuestionStatus,
  QuestionStatusKeys,
  Role,
  UpdateQuestionParams,
  UpdateQuestionResponse,
} from '@koh/common';
import {
  BadRequestException,
  Body,
  ClassSerializerInterceptor,
  Controller,
  Get,
  HttpException,
  HttpStatus,
  NotFoundException,
  Param,
  Patch,
  Post,
  UnauthorizedException,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';
import { Connection, In } from 'typeorm';
import { JwtAuthGuard } from '../guards/jwt-auth.guard';
import {
  NotificationService,
  NotifMsgs,
} from '../notification/notification.service';
import { Roles } from '../decorators/roles.decorator';
import { UserCourseModel } from '../profile/user-course.entity';
import { User, UserId } from '../decorators/user.decorator';
import { UserModel } from '../profile/user.entity';
import { QueueModel } from '../queue/queue.entity';
import { QuestionGroupModel } from './question-group.entity';
import { QuestionRolesGuard } from '../guards/question-role.guard';
import { QuestionModel } from './question.entity';
import { QuestionService } from './question.service';

// NOTE: FIXME: EVERY REQUEST INTO QUESTIONCONTROLLER REQUIRES THE BODY TO HAVE A
// FIELD questionId OR queueId! If not, stupid weird untraceable bugs will happen
// and you will lose a lot of development time
@Controller('questions')
@UseGuards(JwtAuthGuard, QuestionRolesGuard)
@UseInterceptors(ClassSerializerInterceptor)
export class QuestionController {
  constructor(
    private connection: Connection,
    private notifService: NotificationService,
    private questionService: QuestionService,
  ) {}

  @Get(':questionId')
  async getQuestion(
    @Param('questionId') questionId: number,
  ): Promise<GetQuestionResponse> {
    const question = await QuestionModel.findOne(questionId, {
      relations: ['creator', 'taHelped'],
    });

    if (question === undefined) {
      throw new NotFoundException();
    }
    return question;
  }

  @Post()
  @Roles(Role.STUDENT)
  async createQuestion(
    @Body() body: CreateQuestionParams,
    @User() user: UserModel,
  ): Promise<CreateQuestionResponse> {
    const { text, questionType, groupable, queueId, force } = body;

    const queue = await QueueModel.findOne({
      where: { id: queueId },
      relations: ['staffList'],
    });

    if (!queue) {
      throw new NotFoundException(
        ERROR_MESSAGES.questionController.createQuestion.invalidQueue,
      );
    }

    if (!queue.allowQuestions) {
      throw new BadRequestException(
        ERROR_MESSAGES.questionController.createQuestion.noNewQuestions,
      );
    }
    if (!(await queue.checkIsOpen())) {
      throw new BadRequestException(
        ERROR_MESSAGES.questionController.createQuestion.closedQueue,
      );
    }

    const previousUserQuestion = await QuestionModel.findOne({
      where: {
        creatorId: user.id,
        status: In(Object.values(OpenQuestionStatus)),
      },
    });

    if (!!previousUserQuestion) {
      if (force) {
        previousUserQuestion.status = ClosedQuestionStatus.ConfirmedDeleted;
        await previousUserQuestion.save();
      } else {
        throw new BadRequestException(
          ERROR_MESSAGES.questionController.createQuestion.oneQuestionAtATime,
        );
      }
    }

    try {
      const question = await QuestionModel.create({
        queueId: queueId,
        creator: user,
        text,
        questionType,
        groupable,
        status: QuestionStatusKeys.Drafting,
        createdAt: new Date(),
      }).save();
      return question;
    } catch (err) {
      console.error(err);
      throw new HttpException(
        ERROR_MESSAGES.questionController.saveQError,
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Patch(':questionId')
  @Roles(Role.STUDENT, Role.TA, Role.PROFESSOR)
  // TODO: Use queueRole decorator, but we need to fix its performance first
  async updateQuestion(
    @Param('questionId') questionId: number,
    @Body() body: UpdateQuestionParams,
    @UserId() userId: number,
  ): Promise<UpdateQuestionResponse> {
    let question = await QuestionModel.findOne({
      where: { id: questionId },
      relations: ['creator', 'queue', 'taHelped'],
    });
    if (question === undefined) {
      throw new NotFoundException();
    }

    const isCreator = userId === question.creatorId;

    if (isCreator) {
      // Fail if student tries an invalid status change
      if (body.status && !question.changeStatus(body.status, Role.STUDENT)) {
        throw new UnauthorizedException(
          ERROR_MESSAGES.questionController.updateQuestion.fsmViolation(
            'Student',
            question.status,
            body.status,
          ),
        );
      }
      question = Object.assign(question, body);
      try {
        await question.save();
      } catch (err) {
        console.error(err);
        throw new HttpException(
          ERROR_MESSAGES.questionController.saveQError,
          HttpStatus.INTERNAL_SERVER_ERROR,
        );
      }
      return question;
    }

    // If not creator, check if user is TA/PROF of course of question
    const isTaOrProf =
      (await UserCourseModel.count({
        where: {
          userId,
          courseId: question.queue.courseId,
          role: In([Role.TA, Role.PROFESSOR]),
        },
      })) > 0;

    if (isTaOrProf) {
      if (Object.keys(body).length !== 1 || Object.keys(body)[0] !== 'status') {
        throw new UnauthorizedException(
          ERROR_MESSAGES.questionController.updateQuestion.taOnlyEditQuestionStatus,
        );
      }
      await this.questionService.validateNotHelpingOther(body.status, userId);
      return await this.questionService.changeStatus(
        body.status,
        question,
        userId,
      );
    } else {
      throw new UnauthorizedException(
        ERROR_MESSAGES.questionController.updateQuestion.loginUserCantEdit,
      );
    }
  }

  @Post(':questionId/notify')
  @Roles(Role.TA, Role.PROFESSOR)
  async notify(@Param('questionId') questionId: number): Promise<void> {
    const question = await QuestionModel.findOne(questionId, {
      relations: ['queue'],
    });

    if (question === undefined || question === null) {
      throw new HttpException(
        ERROR_MESSAGES.questionController.notFound,
        HttpStatus.NOT_FOUND,
      );
    }

    if (question.status === LimboQuestionStatus.CantFind) {
      try {
        await this.notifService.notifyUser(
          question.creatorId,
          NotifMsgs.queue.ALERT_BUTTON,
        );
      } catch (err) {
        console.error(err);
        throw new HttpException(
          ERROR_MESSAGES.questionController.unableToNotifyUser,
          HttpStatus.INTERNAL_SERVER_ERROR,
        );
      }
    } else if (question.status === LimboQuestionStatus.TADeleted) {
      try {
        await this.notifService.notifyUser(
          question.creatorId,
          NotifMsgs.queue.REMOVED,
        );
      } catch (err) {
        console.error(err);
        throw new HttpException(
          ERROR_MESSAGES.questionController.unableToNotifyUser,
          HttpStatus.INTERNAL_SERVER_ERROR,
        );
      }
    }
  }

  @Post('group')
  @Roles(Role.TA, Role.PROFESSOR)
  async groupQuestions(
    @Body() body: GroupQuestionsParams,
    @UserId() instructorId: number,
  ): Promise<void> {
    const questions = await QuestionModel.find({
      where: {
        id: In(body.questionIds),
      },
      relations: ['taHelped', 'creator'],
    });

    if (!questions.every((q) => q.groupable)) {
      throw new BadRequestException(
        ERROR_MESSAGES.questionController.groupQuestions.notGroupable,
      );
    }

    await this.questionService.validateNotHelpingOther(
      QuestionStatusKeys.Helping,
      instructorId,
    );

    for (const question of questions) {
      await this.questionService.changeStatus(
        QuestionStatusKeys.Helping,
        question,
        instructorId,
      );
    }

    const queue = await QueueModel.findOne({
      where: {
        id: body.queueId,
      },
    });

    const creatorUserCourse = await UserCourseModel.findOne({
      where: {
        courseId: queue.courseId,
        userId: instructorId,
      },
    });

    await QuestionGroupModel.create({
      creatorId: creatorUserCourse.id, // this should be usercourse id
      queueId: body.queueId,
      questions: questions,
    }).save();

    return;
  }

  @Patch('resolveGroup/:group_id')
  @Roles(Role.TA, Role.PROFESSOR)
  async resolveGroup(
    @Param('group_id') groupId: number,
    @UserId() instructorId: number,
  ): Promise<void> {
    const group = await QuestionGroupModel.findOne({
      where: {
        id: groupId,
      },
      relations: ['questions', 'questions.taHelped', 'questions.creator'],
    });

    for (const question of group.questions) {
      // only resolve q's that weren't requeued/can't find
      if (question.status === OpenQuestionStatus.Helping) {
        await this.questionService.changeStatus(
          QuestionStatusKeys.Resolved,
          question,
          instructorId,
        );
      }
    }

    return;
  }
}
