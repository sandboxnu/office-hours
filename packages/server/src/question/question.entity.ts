import { QuestionStatus, QuestionType, Role, StatusInQueue } from '@koh/common';
import { Exclude } from 'class-transformer';
import {
  BaseEntity,
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  SelectQueryBuilder,
} from 'typeorm';
import { UserModel } from '../profile/user.entity';
import { QueueModel } from '../queue/queue.entity';
import { canChangeQuestionStatus } from './question-fsm';
import { QuestionGroupModel } from './question-group.entity';

@Entity('question_model')
export class QuestionModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @ManyToOne((type) => QueueModel, (q) => q.questions)
  @JoinColumn({ name: 'queueId' })
  @Exclude()
  queue: QueueModel;

  @Column({ nullable: true })
  @Exclude()
  queueId: number;

  @Column('text')
  text: string;

  @ManyToOne((type) => UserModel)
  @JoinColumn({ name: 'creatorId' })
  creator: UserModel;

  @Column({ nullable: true })
  @Exclude()
  creatorId: number;

  @ManyToOne((type) => UserModel)
  @JoinColumn({ name: 'taHelpedId' })
  taHelped: UserModel;

  @Column({ nullable: true })
  @Exclude()
  taHelpedId: number;

  @Column()
  createdAt: Date;

  // When the question was first helped (doesn't overwrite)
  @Column({ nullable: true })
  @Exclude()
  firstHelpedAt: Date;

  // When the question was last helped (getting help again on priority queue overwrites)
  @Column({ nullable: true })
  helpedAt: Date;

  // When the question leaves the queue
  @Column({ nullable: true })
  closedAt: Date;

  @Column('text', { nullable: true })
  questionType: QuestionType;

  @Column('text')
  status: QuestionStatus;

  @Column({ nullable: true })
  location: string;

  @Column()
  groupable: boolean;

  @ManyToOne((type) => QuestionGroupModel, { nullable: true })
  @JoinColumn({ name: 'groupId' })
  group: QuestionGroupModel;

  @Column({ nullable: true })
  @Exclude()
  groupId: number;

  /**
   * Change the status of the question as the given role
   *
   * @returns whether status change succeeded
   */
  public changeStatus(newStatus: QuestionStatus, role: Role): boolean {
    if (canChangeQuestionStatus(this.status, newStatus, role)) {
      this.status = newStatus;
      return true;
    } else {
      return false;
    }
  }

  /**
   * Scopes
   */
  static inQueueWithStatus(
    queueId: number,
    statuses: QuestionStatus[],
  ): SelectQueryBuilder<QuestionModel> {
    return this.createQueryBuilder('question')
      .where('question.queueId = :queueId', { queueId })
      .andWhere('question.status IN (:...statuses)', {
        statuses,
      })
      .orderBy('question.createdAt', 'ASC');
  }

  /**
   * Questions that are open in the queue (not in priority queue)
   */
  static waitingInQueue(queueId: number): SelectQueryBuilder<QuestionModel> {
    return QuestionModel.inQueueWithStatus(queueId, StatusInQueue);
  }
}
