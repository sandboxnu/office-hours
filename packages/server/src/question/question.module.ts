import { Module } from '@nestjs/common';
import { NotificationModule } from '../notification/notification.module';
import { QuestionController } from './question.controller';
import { QuestionSubscriber } from './question.subscriber';
import { QueueModule } from '../queue/queue.module';

@Module({
  controllers: [QuestionController],
  providers: [QuestionSubscriber],
  imports: [NotificationModule, QueueModule],
})
export class QuestionModule {}
