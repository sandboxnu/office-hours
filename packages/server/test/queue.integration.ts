import { OpenQuestionStatus, Role } from '@koh/common';
import { QuestionModel } from 'question/question.entity';
import { QueueModule } from '../src/queue/queue.module';
import {
  CourseFactory,
  QuestionFactory,
  QueueFactory,
  StudentCourseFactory,
  TACourseFactory,
  UserCourseFactory,
  UserFactory,
} from './util/factories';
import { setupIntegrationTest } from './util/testUtils';
import { QueueModel } from '../src/queue/queue.entity';

async function delay(ms) {
  // return await for better async stack trace support in case of errors.
  return await new Promise((resolve) => setTimeout(resolve, ms));
}

describe('Queue Integration', () => {
  const supertest = setupIntegrationTest(QueueModule);

  describe('GET /queues/:id', () => {
    it('get a queue', async () => {
      const course = await CourseFactory.create();
      const ta = await UserFactory.create();
      await TACourseFactory.create({ course: course, user: ta });

      const queue = await QueueFactory.create({
        courseId: course.id,
        course: course,
        questions: [await QuestionFactory.create()],
        staffList: [ta],
      });
      const userCourse = await UserCourseFactory.create({
        user: await UserFactory.create(),
        course: queue.course,
      });

      const res = await supertest({ userId: userCourse.user.id })
        .get(`/queues/${queue.id}`)
        .expect(200);
      expect(res.body).toMatchObject({
        id: 2,
        notes: null,
        queueSize: 1,
        room: 'Online',
        staffList: expect.any(Array),
        isOpen: true,
      });
    });

    it('is not open when there are no TAs present', async () => {
      const queue = await QueueFactory.create({});
      const userCourse = await UserCourseFactory.create({
        user: await UserFactory.create(),
        course: queue.course,
      });

      const res = await supertest({ userId: userCourse.user.id })
        .get(`/queues/${queue.id}`)
        .expect(200);
      expect(res.body).toMatchObject({
        // isOpen: false,
        isOpen: true,
      });
    });

    it('is open when there are TAs present', async () => {
      const course = await CourseFactory.create();
      const ta = await UserFactory.create();
      await TACourseFactory.create({ course: course, user: ta });
      const queue = await QueueFactory.create({
        course: course,
        staffList: [ta],
      });

      const userCourse = await UserCourseFactory.create({
        user: await UserFactory.create(),
        course: queue.course,
      });

      const res = await supertest({ userId: userCourse.user.id })
        .get(`/queues/${queue.id}`)
        .expect(200);
      expect(res.body).toMatchObject({
        isOpen: true,
      });
    });

    it('returns 404 on non-existent course', async () => {
      const course = await CourseFactory.create();
      const ta = await UserFactory.create();
      await TACourseFactory.create({ course: course, user: ta });

      const queue = await QueueFactory.create({ course, staffList: [ta] });
      const user = await UserFactory.create();

      await supertest({ userId: user.id })
        .get(`/queues/${queue.id + 999}`)
        .expect(404);
    });

    it('returns 401 when not logged in', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({
        course: course,
        questions: [await QuestionFactory.create()],
      });

      const res = await supertest({ userId: 99 })
        .get(`/queues/${queue.id}`)
        .expect(401);
      expect(res.body).toMatchSnapshot();
    });

    it('returns 404 when user is not in course', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({
        courseId: course.id,
        course: course,
        questions: [await QuestionFactory.create()],
      });
      const userCourse = await UserCourseFactory.create({
        user: await UserFactory.create(),
      });

      await supertest({ userId: userCourse.user.id })
        .get(`/queues/${queue.id}`)
        .expect(404);
    });
  });

  describe('GET /queues/:id/questions', () => {
    it('returns questions in the queue', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({
        course: course,
        questions: [
          await QuestionFactory.create({
            text: 'in queue',
            createdAt: new Date('2020-03-01T05:00:00.000Z'),
          }),
        ],
      });
      await QuestionFactory.create({
        text: 'not in queue',
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });
      const userCourse = await UserCourseFactory.create({
        user: await UserFactory.create(),
        course: queue.course,
        courseId: queue.course.id,
      });

      const res = await supertest({ userId: userCourse.user.id })
        .get(`/queues/${queue.id}/questions`)
        .expect(200);

      expect(res.body).toMatchSnapshot();
      expect(res.body.queue[0].creator).not.toHaveProperty('firstName');
      expect(res.body.queue[0].creator).not.toHaveProperty('lastName');
    });

    it('student can see their own questions private data', async () => {
      const course = await CourseFactory.create();
      const userCourse = await UserCourseFactory.create({
        user: await UserFactory.create(),
        course: course,
        courseId: course.id,
      });
      const queue = await QueueFactory.create({
        course: course,
        questions: [
          await QuestionFactory.create({
            text: 'in queue',
            createdAt: new Date('2020-03-01T05:00:00.000Z'),
            creator: userCourse.user,
            creatorId: userCourse.userId,
          }),
        ],
      });
      await QuestionFactory.create({
        text: 'not in queue',
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });

      const res = await supertest({ userId: userCourse.user.id })
        .get(`/queues/${queue.id}/questions`)
        .expect(200);
      expect(res.body.queue[0].creator).toHaveProperty('firstName');
      expect(res.body.queue[0].creator).toHaveProperty('lastName');
    });

    it('returns all creator data for ta', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({
        course: course,
        questions: [
          await QuestionFactory.create({
            text: 'in queue',
            createdAt: new Date('2020-03-01T05:00:00.000Z'),
          }),
        ],
      });
      await QuestionFactory.create({
        text: 'not in queue',
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });
      const ta = await TACourseFactory.create({
        user: await UserFactory.create(),
        course: queue.course,
        courseId: queue.course.id,
      });

      const res = await supertest({ userId: ta.user.id })
        .get(`/queues/${queue.id}/questions`)
        .expect(200);
      expect(res.body.queue[0].creator).toHaveProperty('firstName');
      expect(res.body.queue[0].creator).toHaveProperty('lastName');
    });

    it('returns 404 when a user is not a member of the course', async () => {
      const course = await CourseFactory.create();
      const queue = await QueueFactory.create({
        course: course,
        questions: [
          await QuestionFactory.create({
            text: 'in queue',
            createdAt: new Date('2020-03-01T05:00:00.000Z'),
          }),
        ],
      });
      await QuestionFactory.create({
        text: 'not in queue',
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });
      const userCourse = await UserCourseFactory.create({
        user: await UserFactory.create(),
      });

      await supertest({ userId: userCourse.user.id })
        .get(`/queues/${queue.id}/questions`)
        .expect(404);
    });

    it('returns 404 when queue does not exist', async () => {
      const user = await UserFactory.create();

      await supertest({ userId: user.id })
        .get(`/queues/8291390/questions`)
        .expect(404);
    });
  });

  describe('POST /queues/:id/clean', () => {
    // Create a queue that is closed and has a question ready to be cleaned
    const cleanableQueue = async () => {
      const queue = await QueueFactory.create({
        room: 'The Alamo',
      });
      await QuestionFactory.create({ queue: queue });
      return queue;
    };
    it('cleans the queue', async () => {
      const queue = await cleanableQueue();
      const tcf = await TACourseFactory.create({
        course: queue.course,
        user: await UserFactory.create(),
      });

      expect(
        await QuestionModel.inQueueWithStatus(
          queue.id,
          Object.values(OpenQuestionStatus),
        ).getCount(),
      ).toEqual(1);

      await supertest({ userId: tcf.userId })
        .post(`/queues/${queue.id}/clean`)
        .expect(201);

      await delay(100);
      expect(
        await QuestionModel.inQueueWithStatus(
          queue.id,
          Object.values(OpenQuestionStatus),
        ).getCount(),
      ).toEqual(0);
    });

    it('does not allow students access', async () => {
      const queue = await cleanableQueue();
      const scf = await StudentCourseFactory.create({
        course: queue.course,
        user: await UserFactory.create(),
      });

      await supertest({ userId: scf.userId })
        .post(`/queues/${queue.id}/clean`)
        .expect(401);

      await delay(100);
      /// questions should still be there
      expect(
        await QuestionModel.inQueueWithStatus(
          queue.id,
          Object.values(OpenQuestionStatus),
        ).getCount(),
      ).toEqual(1);
    });
  });

  describe('DELETE /queues/:id', () => {
    it('disables queues on hit', async () => {
      const course = await CourseFactory.create();
      const ta = await TACourseFactory.create({
        course: course,
        user: await UserFactory.create(),
      });
      const queue = await QueueFactory.create({
        course: course,
      });

      expect(queue.isDisabled).toBeFalsy();

      await supertest({ userId: ta.userId })
        .delete(`/queues/${queue.id}`)
        .expect(200);

      const postQueue = await QueueModel.findOne({ id: queue.id });
      expect(postQueue.isDisabled).toBeTruthy();
    });

    it('doesnt allow students to delete queue', async () => {
      const course = await CourseFactory.create();
      const stu = await StudentCourseFactory.create({
        course: course,
        user: await UserFactory.create(),
      });
      const queue = await QueueFactory.create({
        course: course,
      });

      expect(queue.isDisabled).toBeFalsy();

      await supertest({ userId: stu.userId })
        .delete(`/queues/${queue.id}`)
        .expect(401);

      const postQueue = await QueueModel.findOne({ id: queue.id });
      expect(postQueue.isDisabled).toBeFalsy();
    });

    it('doesnt allow TAs to disable prof queues', async () => {
      const course = await CourseFactory.create();
      const ta = await TACourseFactory.create({
        course: course,
        user: await UserFactory.create(),
      });
      const queue = await QueueFactory.create({
        course: course,
        isProfessorQueue: true,
      });

      expect(queue.isDisabled).toBeFalsy();

      await supertest({ userId: ta.userId })
        .delete(`/queues/${queue.id}`)
        .expect(401);
    });

    it('allows professors to disable prof queues', async () => {
      const course = await CourseFactory.create();
      const prof = await UserCourseFactory.create({
        course: course,
        user: await UserFactory.create(),
        role: Role.PROFESSOR,
      });

      const queue = await QueueFactory.create({ course });

      expect(queue.isDisabled).toBeFalsy();

      await supertest({ userId: prof.userId })
        .delete(`/queues/${queue.id}`)
        .expect(200);

      const postQueue = await QueueModel.findOne({ id: queue.id });
      expect(postQueue.isDisabled).toBeTruthy();
    });
    it('returns 404 on nonexistent queues', async () => {
      const stu = await StudentCourseFactory.create({
        user: await UserFactory.create(),
      });
      await supertest({ userId: stu.userId }).delete(`/queues/998`).expect(404);
    });
  });
});
