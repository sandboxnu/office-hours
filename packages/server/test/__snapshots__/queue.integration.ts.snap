// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queue Integration GET /queues/:id returns 401 when not logged in 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Must be logged in",
  "statusCode": 401,
}
`;

exports[`Queue Integration GET /queues/:id/questions returns questions in the queue 1`] = `
Object {
  "groups": Array [],
  "priorityQueue": Array [],
  "questionsGettingHelp": Array [],
  "queue": Array [
    Object {
      "closedAt": null,
      "createdAt": "2020-03-01T05:00:00.000Z",
      "creator": Object {
        "id": 1,
      },
      "groupable": true,
      "helpedAt": null,
      "id": 1,
      "isOnline": null,
      "location": null,
      "questionType": "Other",
      "status": "Queued",
      "taHelped": null,
      "text": "in queue",
    },
  ],
  "unresolvedAlerts": Array [],
}
`;
